### PR TITLE
Added InstanceReferenceKind to IInstanceReferenceOperation for tellin…

### PR DESCRIPTION
**Pull request by 333fred**
…g between this. usage and initializer usage.

Split out part of https://github.com/dotnet/roslyn/pull/25322 to reduce the resulting size of the PR, and update the base to current features/dataflow. I've addressed initial relevant comments from that review.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred It looks like there are legitimate test failures.

File: src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs]: >This [](start = 117, length = 4)

Did you mean to say "Me" here?

File: src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs]: >Reference to a member [](start = 12, length = 21)

This doesn't sound right. I don't think IInstanceReferenceOperation ever references a member.

File: src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs]: >Reference to the containing type. [](start = 12, length = 33)

"Reference to **an instance of** the containing type." ?

**Comment by CyrusNajmabadi** [comment]
> "and initializer usage."

Can you give an example of what you mean by that?  Thanks!

**Comment by AlekseyTs** [comment]
>     /// Current usage:

It looks like this list here should be expanded with more items

---
Refers to: src/Compilers/Core/Portable/Operations/IInstanceReferenceExpression.cs:8 in 33db5e0. [](commit_id = 33db5e098fbe4821b745356988ee3e00c16a1709, deletion_comment = False)

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: In this file, I would expect to see only changes around IInstanceReferenceOperation.ReferenceKind property and InstanceReferenceKind type. Please revert all other unnecessary changes.

Comment (333fred) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: The codefix that adds things to the public api automatically sorts this file. If we don't do this now, we'll have to manually fix this file up every time we want to add anything new. I'm going to leave this as is.

---
In reply to: [180511221](https://github.com/dotnet/roslyn/pull/26055#discussion_r180511221) [](ancestors = 180511221)

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: If the sorting is necessary, we can apply it in a separate targeted PR.

---
In reply to: [180514762](https://github.com/dotnet/roslyn/pull/26055#discussion_r180514762) [](ancestors = 180514762,180511221)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >ContainingTypeInstance [](start = 72, length = 22)

Is this the right value? Should this be ```InitializedInstance``` instead?

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >InitializedInstance [](start = 79, length = 19)

I think this could be an implicit receiver in an With statement. We probably should have a special kind for this case and use the appropriate one here.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >InitializedInstance [](start = 79, length = 19)

I think this could be an implicit receiver in an With statement. We probably should have a special kind for this case and use the appropriate one here.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IWithStatement.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IWithStatement.vb]: >InitializedInstance [](start = 62, length = 19)

There is no initialized instance in this scenario

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: >VisitInstanceReference [](start = 29, length = 22)

Consider making an adjustment to TestOperationVisitor as well. At least we should verify that the new property doesn't throw.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1). Tests are not looked at.

**Comment by 333fred** [comment]
This is object/collection initializers, or With statements. See the IOperationTests_IObjectCreationExpression.cs tests for a C# example.

---
In reply to: [380186314](https://github.com/dotnet/roslyn/pull/26055#issuecomment-380186314) [](ancestors = 380186314)

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Addressing this by renaming `InitializedInstance` to `ImplicitReceiver`. If we later get a request to be able to tell whether the instance is actually being initialized or just used as part of a With statement, then we can revisit or add some other boolean property.

---
In reply to: [180531576](https://github.com/dotnet/roslyn/pull/26055#discussion_r180531576) [](ancestors = 180531576)

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt
Comment (333fred) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Given that this was a generated change, I'm going to leave it as is. I did not manually change this file at all, this is purely codefix.

---
In reply to: [180522270](https://github.com/dotnet/roslyn/pull/26055#discussion_r180522270) [](ancestors = 180522270,180515588,180514762,180511221)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [380211901](https://github.com/dotnet/roslyn/pull/26055#issuecomment-380211901) [](ancestors = 380211901)

**Comment by 333fred** [comment]
Test failure is https://github.com/dotnet/roslyn/issues/12106

File: src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs]: >in a C# object or collection initializer [](start = 54, length = 40)

VB also has these.

File: src/Compilers/Core/Portable/Operations/IInstanceReferenceExpression.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/IInstanceReferenceExpression.cs]: >C# object or collection initializers. [](start = 13, length = 37)

Why C# only? VB also has these.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >InstanceReferenceKind.ImplicitReceiver [](start = 50, length = 38)

Could you please point me to IOperation tests that go through this code path?

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 2)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Some of the tests I found:
* ObjectCreationWithArrayInitializer
* ObjectCreationWithInvalidMemberInitializer
* ObjectCreationWithInvalidInitializer
* DeconstructDelegate2
* ObjectInitializerTest_IncompleteComplexElementInitializerExpression
* CS0246_ERR_SingleTypeNameNotFound
* CS0117ERR_NoSuchMember
* AddPropertyUsedLikeMethod

---
In reply to: [180823954](https://github.com/dotnet/roslyn/pull/26055#discussion_r180823954) [](ancestors = 180823954)

**Comment by 333fred** [comment]
Ping @dotnet/roslyn-compiler @mavasani for reviews.

**Comment by 333fred** [comment]
Ping @dotnet/roslyn-compiler @mavasani for reviews.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs]: Thanks, this is very useful information. Do we have an issue to track exposing a helper extension method that maps an `IInstanceReferenceOperation` with an `ImplicitReceiver` to the receiver operation? That will be a useful helper for dataflow analysis visitor, where when you encounter an `IInstanceReferenceOperation`, you normally want to return the tracked state/data for the receiver being referenced.

**Review by mavasani — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs]: Not currently, can you file one?

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs]: Filed https://github.com/dotnet/roslyn/issues/26152

**Comment by 333fred** [comment]
Ping @dotnet/roslyn-compiler for another review please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_unit64_prtest please

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review. We need a second review for this PR, which **11 days old**.

**Review by jcouv — APPROVED** [review]
LGTM Thanks

**Comment by 333fred** [comment]
Test failures are https://github.com/dotnet/roslyn/issues/26037, the fix for which isn't in this branch yet.

**Pull request merged**
